### DISCUSSION TITLE
Idiomatic way to create a read timeout on socket

### DISCUSSION THREAD

**Discussion by DariuszOstolski** [tokio-rs/tokio#3793] [Q&A]
Hi,

Maybe a stupid question but what is the most idiomatic way of creating a read socket timeout?

Currently I have following code:
`async fn handle_client(mut client: Client) {
        debug!("New client connected: {}", client.addr);
        let handler = handler::BinaryHandler::new(client.store.clone());

        // Here for every packet we get back from the `Framed` decoder,
        // we parse the request, and if it's valid we generate a response
        // based on the values in the storage.
        loop {
            match timeout(
                Duration::from_secs(client.config.rx_timeout_secs as u64),
                client.stream.read_frame(),
            )
            .await
            {
....
`
Is it the best idiomatic way to write it?

BR,
Dariusz

**Comment by Darksonn** [comment]
That's a fine way to do it. You can also achieve the same with a `tokio::select!`, but there isn't much reason to do that if all you need is a timeout.

**Reply by ririsoft** [reply]
@Darksonn what would your recommend for applying read (or write) timeout within a `io::copy` operation please ? `tokio::select` or `tokio::io::timeout` does not help here because I only want the timeout to occure if no read or write happened within a given period of time during the copy operation. This is particularly useful for network operations for instance. So far I found the [tokio-io-timeout](https://docs.rs/tokio-io-timeout) crate that does the job, but I was wondering if there is a more straightforward way to do this with tokio only.

**Reply by Darksonn** [reply]
There isn't an easy way to embed it inside `io::copy`. You could use a crate like the one you linked, or replace the copy with some sort of loop.
